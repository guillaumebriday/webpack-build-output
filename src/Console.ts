import chalk from "chalk"
import readline from "readline"
import { WriteStream } from "tty"

export interface ConsoleOutput {
  get width(): number
  get height(): number

  clear(): void
  heading(text: string): void
  success(text: string): void
  write(str: string): void
}

export class Console {
  constructor(private stream: WriteStream) {}

  public get width() {
    return this.stream.columns
  }

  public get height() {
    return this.stream.rows
  }

  public clear() {
    this.write("\n".repeat(this.height))
    readline.cursorTo(this.stream, 0, 0)
    readline.clearScreenDown(this.stream)
  }

  /**
   * Print a block section heading.
   */
  public heading(text: string) {
    this.write("\n")
    this.write(chalk.bgBlue.white.bold(this.section(text)))
    this.write("\n")
  }

  /**
   * Print a success message
   */
  public success(text: string) {
    this.write(chalk.green.bold(`âœ” ${text}`))
    this.write("\n")
  }

  /**
   * Write directly to the console
   */
  public write(str: string) {
    this.stream.write(str)
  }

  /**
   * Create a block section.
   */
  private section(text: string) {
    const padLength = 3
    const padding = " ".repeat(padLength)

    text = `${padding}${text}${padding}`

    const line = " ".repeat(text.length)

    return `${line}\n${text}\n${line}`
  }
}
