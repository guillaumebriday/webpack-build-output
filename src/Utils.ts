import stripAnsi from "strip-ansi"

/**
 * @param {number} size the size in bytes
 * @returns {string} the formatted size
 */
export function formatSize(size: number): string {
  if (Number.isNaN(size) === true) {
    return "unknown"
  }

  const abbreviations = ["bytes", "KB", "MB", "GB", "TB"]

  const index = Math.floor(Math.log(size) / Math.log(10 ** 3))
  const amount = size / 1024 ** index

  return `${amount.toPrecision(3)} ${abbreviations[index]}`
}

let patched = false

export function monkeyPatchTruncate() {
  if (patched) {
    return
  }

  patched = true

  const utils = require("cli-table3/src/utils")
  const oldTruncate = utils.truncate

  // TODO:
  // We want to replace this truncation routine with something
  // custom so we can control how and where truncation happens
  utils.truncate = (str: string, desiredLength: number, truncateChar: string) => {
    if (stripAnsi(str).length > desiredLength) {
      str = `â€¦${str.substr(-desiredLength + 2)}`
    }

    return oldTruncate(str, desiredLength, truncateChar)
  }
}
