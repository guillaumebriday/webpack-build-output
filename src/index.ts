export * from "./Options"
export * from "./BuildOutputPlugin"
