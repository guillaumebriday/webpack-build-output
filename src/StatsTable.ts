import chalk from "chalk"
import Table from "cli-table3"
import stripAnsi from "strip-ansi"
import { StatsCompilation } from "webpack"
import { formatSize, monkeyPatchTruncate } from "./Utils"
import { Options } from "./Options"
import { ConsoleOutput } from "./Console"

export class StatsTable {
  constructor(private options: Options, private console: ConsoleOutput) {}

  /**
   * Print the stats table to the screen
   */
  public render(data: StatsCompilation): string {
    const table = new Table({
      head: [chalk.bold("File"), chalk.bold("Size")],
      colWidths: [35],
      colAligns: ["right"],
      style: {
        head: [],
        compact: true,
      },
    })

    const assets = this.sortAssets(data)

    for (const asset of assets) {
      table.push([chalk.green(asset.name), formatSize(asset.size)])
    }

    this.extendTableWidth(table, this.options.width ?? this.console.width, this.options.maxWidth ?? Infinity)

    monkeyPatchTruncate()

    return table.toString()
  }

  private sortAssets(data: StatsCompilation) {
    let assets = data.assets!.flatMap((asset) => [asset, ...(Array.isArray(asset.related) ? asset.related : [])])

    assets = assets.sort((a, b) => {
      return a.name.localeCompare(b.name) || a.size - b.size
    })

    return assets
  }

  /**
   * Extend the width of the table
   *
   * Currently only increases the file column size
   */
  private extendTableWidth(table: Table.Table, targetWidth: number, maxWidth: number) {
    if (!targetWidth) {
      return
    }

    const tableWidth = this.calculateWidth(table)
    const fileColIncrease = Math.min(targetWidth - tableWidth, maxWidth - tableWidth)

    if (fileColIncrease <= 0) {
      return
    }

    // @ts-ignore
    table.options.colWidths[0] += fileColIncrease
  }

  /**
   * Calculate the width of the CLI Table
   *
   * `table.width` does not report the correct width
   * because it includes ANSI control characters
   */
  private calculateWidth(table: Table.Table) {
    const firstRow = table.toString().split("\n")[0]

    return stripAnsi(firstRow).length
  }
}
