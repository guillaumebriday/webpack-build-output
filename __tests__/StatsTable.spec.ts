import colors from "colors/safe"
import { StatsAsset } from "webpack"

import { StatsTable } from "../src/StatsTable"
import { FakeConsole } from "./FakeConsole"

beforeEach(() => {
  colors.disable()
  process.env.FORCE_COLOR = "0"
})

afterEach(() => {
  delete process.env.FORCE_COLOR
})

test("The build output table renders as expected", () => {
  const table = new StatsTable({ showRelated: true }, new FakeConsole(80, 200))

  const output = table.render({
    assets: [
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/app.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/js/bar.js",
        size: 1200,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/extracted.css",
        size: 1500,
      }),
      asset({
        name: "foo/foo/foo/foo/foo/foo/foo/css/app.css",
        size: 1000,
      }),
    ],
  })

  expect(`\n${output}\n`).toMatchSnapshot()
})

function asset(asset: Partial<StatsAsset>): StatsAsset {
  return {
    name: "foo/foo/foo/foo/foo/foo/foo/js/app.js",
    size: 1200,
    related: [],
    type: "test",
    info: {},
    emitted: false,
    comparedForEmit: false,
    cached: false,
    ...asset,
  }
}
