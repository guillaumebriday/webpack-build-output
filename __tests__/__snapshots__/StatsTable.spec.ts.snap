// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The build output table renders as expected 1`] = `
"
┌───────────────────────────────────────────────────────────────────┬──────────┐
│                                                              File │ Size     │
├───────────────────────────────────────────────────────────────────┼──────────┤
│                           foo/foo/foo/foo/foo/foo/foo/css/app.css │ 0.977 KB │
│                     foo/foo/foo/foo/foo/foo/foo/css/extracted.css │ 1.46 KB  │
│                             foo/foo/foo/foo/foo/foo/foo/js/app.js │ 1.17 KB  │
│                             foo/foo/foo/foo/foo/foo/foo/js/bar.js │ 1.17 KB  │
└───────────────────────────────────────────────────────────────────┴──────────┘
"
`;
